csrc/multidevice/device_mesh.h

```diff
@@ -31,23 +30,41 @@ class DeviceMesh final {
   // There are no such contention for std::initializer_list so I chose to
   // allow implicit conversion for that. This allows users to write `DeviceMesh
   // mesh = {1, 2};`, which is more concise.
-  explicit DeviceMesh(std::vector<DeviceIdxType> devices = {});
+  // When no shape is specified, a 1D DeviceMesh is created by default.
+  explicit DeviceMesh(
+      std::vector<DeviceIdxType> devices = {},
+      std::vector<int64_t> shape = {});
   DeviceMesh(std::initializer_list<DeviceIdxType> devices);
   DeviceMesh(const DeviceMesh&) = default;
   DeviceMesh(DeviceMesh&&) = default;
   DeviceMesh& operator=(const DeviceMesh&) = default;
   DeviceMesh& operator=(DeviceMesh&&) = default;
 
-  // Creates a device mesh of [0 .. num_devices-1]. I didn't make it a
+  // Creates a device mesh of [0 ... num_devices-1]. I didn't make it a
   // constructor because single-element initializer lists would be directed to
   // use that instead of the constructor for vectors.
   static DeviceMesh createForNumDevices(int64_t num_devices);
 
-  // Returns the number of devices in the mesh
+  // Creates a device mesh with the specified shape made of devices from
+  // [0 ... size], where size is the total number of devices specified by the
+  // shape.
+  static DeviceMesh createForShape(std::vector<int64_t> shape);
+
+  // Returns the total number of devices in the mesh
   int64_t size() const {
     return static_cast<int64_t>(vector_.size());
   }
 
+  // Returns size of an axis in the mesh
+  int64_t size(int64_t axis) {
+    return shape_.at(axis);
+  }
+
+  // Returns the shape of the device mesh
+  const std::vector<int64_t>& shape() const {
+    return shape_;
+  }
+
   // Returns a vector containing the device indices of the mesh
   const std::vector<DeviceIdxType>& vector() const {
     return vector_;
```

Comment:
Unless there is a reason to keep this, I think we can get rid of this function.

Reply:
agreed

Reply:
actually imo shape() is enough... up to you